src/DI/ContainerBuilder.php

```diff
@@ -178,17 +178,22 @@ public function getByType($class)
 
 
 	/**
-	 * Gets the service names of the specified type.
-	 * @param string
-	 * @return string[]
+	 * Gets the service names and definitions of the specified type.
+	 * @param  string
+	 * @return ServiceDefinition[]
 	 */
-	public function findByType($class, $autowired = TRUE)
+	public function findByType($class)
 	{
-		self::checkCase($class);
-		return array_merge(
-			isset($this->classes[$class][TRUE]) ? $this->classes[$class][TRUE] : array(),
-			!$autowired && isset($this->classes[$class][FALSE]) ? $this->classes[$class][FALSE] : array()
-		);
+		$is_subclass_of = PHP_VERSION_ID >= 50307 ? 'is_subclass_of' : function($class, $of) {
+			return array_key_exists($of, class_parents($class) + class_implements($class));
+		};
+		$found = array();
+		foreach ($this->definitions as $name => $def) {
```

Comment:
I'm a bit concerned about the performance of this method. Also, it still depends on the `prepareClassList`, because it can't resolve types from `@return` annotations by itself.